# Capstone projet week 3: Segmenting and Clustering Neighborhoods in Toronto

# First part

In this first part I obtain the list of neighbourhoods according to the instructions 

In [1]:
import pandas as pd
!pip install lxml

In [2]:
wiki_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_table = pd.read_html(wiki_url)    

The previous operation gives a list of three elements, of which the fist one, index [0] is the table I am interested in. I take the first element and turn it into a dataframe. 

In [3]:
wiki_df = pd.DataFrame(wiki_table[0])
wiki_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [4]:
wiki_df.size

540

In [5]:
wiki_df.drop(wiki_df[wiki_df.Borough == 'Not assigned'].index, inplace = True)

In [6]:
wiki_df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In the next command I reindex the rows of the df

In [7]:
wiki_df.reset_index(drop = True, inplace = True)

In the current version of the Wikipedia page the neighbourhoods are already aggregated by postal code, so no further aggregation is necessary. 

There are no lines with no assigned neighbourhood, so no further correction is needed

In [8]:
wiki_df[wiki_df.Neighbourhood == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [9]:
wiki_df.shape

(103, 3)

There are 103 rows and three columns in the dataframe

# Second part

In the second part I fetch the list of coordinates for each neighbourhood and I match the postal codes with the coordinates

In [10]:
csv_lat_long = pd.read_csv('http://cocl.us/Geospatial_data')

In [11]:
csv_lat_long.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [12]:
lat_new = pd.Series([],dtype = 'float64') 
long_new = pd.Series([],dtype = 'float64')

In [13]:
for counter, PC in enumerate(wiki_df.loc[:,'Postal Code']):
                    #print(PC)
                    lat_new[counter] = csv_lat_long.loc[csv_lat_long['Postal Code'] == PC,'Latitude'].tolist()[0]
                    #print(lat_new[counter])
                    long_new[counter] = csv_lat_long.loc[csv_lat_long['Postal Code'] == PC,'Longitude'].tolist()[0]
                    #print(long_new[counter])

In [14]:
wiki_df['Latitude'] = lat_new

In [15]:
wiki_df['Longitude'] = long_new

In [16]:
wiki_df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# Third part

In the third part I obtain the most common venues per neighbourhood from FourSquare and I apply a k-mean clustering method to group the neighbourhoods, like we did in the course for the Manhattan neighbourhoods. 

## Importing the libraries

In [17]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


## Creating a map with the Toronto neighbourhoods

Fetching Toronto coordinates

In [18]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


creating the map

In [19]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(wiki_df['Latitude'], wiki_df['Longitude'], wiki_df['Borough'], wiki_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's define the Foursquare access credentials

In [20]:
CLIENT_ID = 'CT2L4SYXIDPGMQCYAWJ1KOC2EBNEISGNBUZ2CWFD5MYQXKJ5' # your Foursquare ID
CLIENT_SECRET = 'GGLIFFLM5CBWUYF3KUITLIQHJRDYHZRGNMUFSDMCM2PYWG1W' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CT2L4SYXIDPGMQCYAWJ1KOC2EBNEISGNBUZ2CWFD5MYQXKJ5
CLIENT_SECRET:GGLIFFLM5CBWUYF3KUITLIQHJRDYHZRGNMUFSDMCM2PYWG1W


Let's define a function that explores the different neighbourhoods in Toronto

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    #counter = 0
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
       # counter = counter + 1
      #  print(counter)
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
radius = 500
LIMIT = 100

Now I use the function to create a new dataframe with the venues for Toronto

In [23]:
toronto_venues = getNearbyVenues(names = wiki_df['Neighbourhood'],
                                   latitudes = wiki_df['Latitude'],
                                   longitudes = wiki_df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [24]:
print(toronto_venues.shape)
toronto_venues.head()

(2156, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1             649 Variety       43.754513       -79.331942  Convenience Store  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

In [25]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                                4   
Alderwood, Long Branch                                                   8   
Bathurst Manor, Wilson Heights, Downsview North                         23   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       27   
Berczy Park                                                             57   
Birch Cliff, Cliffside West                                              4   
Brockton, Parkdale Village, Exhibition Place                            25   
Business reply mail Processing Centre, South Ce...                      19   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   12   
Cedarbrae                                                                8   
Central Bay Street                                                      67   
Christie                                                                16   
Church and Wellesley                                                    77   
Clarks Corners, Tam O'Shanter, Sullivan                                 11   
Cliffside, Cliffcrest, Scarborough Village West                          2   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                                                         7   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                         4   
Don Mills                                                               28   
Dorset Park, Wexford Heights, Scarborough Town ...                       6   
Downsview                                                               13   
Dufferin, Dovercourt Village                                            14   
East Toronto, Broadview North (Old East York)                            4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                       8   
Fairview, Henry Farm, Oriole                                            64   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Glencairn                                                                4   
Golden Mile, Clairlea, Oakridge                                          9   
Guildwood, Morningside, West Hill                                        8   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                           24   
Hillcrest Village                                                        5   
Humber Summit                                                            2   
Humberlea, Emery                                                         2   
Humewood-Cedarvale                                                       4   
India Bazaar, The Beaches West                                          21   
Kennedy Park, Ionview, East Birchmount Park                              6   
Kensington Market, Chinatown, Grange Park                               66   
Kingsview Village, St. Phillips, Martin Grove G...                       3   
Lawrence Manor, Lawrence Heights                                        10   
Lawrence Park                                                            4   
Leaside                                                                 32   
Little Portugal, Trinity                                                46   
Malvern, Rouge                                                           1

Here I group the venues by neighbourhood

In [26]:
toronto_venues.groupby('Neighbourhood').count().shape

(95, 6)

Here I create a dataframe that shows the type of venue for each venue and each neighbourhood

In [27]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2         Parkwoods                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0          0       0           0              0          0         0   
1          0       0           0              0          0         0   
2          0       0           0              0          0         0   
3          0       0           0              0          0         0   
4          0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Pla

In [28]:
toronto_onehot.shape

(2156, 273)

Here I show for each neighbourhood the probability of finding a certain venue, for each venue

In [29]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                           Agincourt              0.000   
1                              Alderwood, Long Branch              0.000   
2     Bathurst Manor, Wilson Heights, Downsview North              0.000   
3                                     Bayview Village              0.000   
4                   Bedford Park, Lawrence Manor East              0.000   
5                                         Berczy Park              0.000   
6                         Birch Cliff, Cliffside West              0.000   
7        Brockton, Parkdale Village, Exhibition Place              0.000   
8   Business reply mail Processing Centre, South C...              0.000   
9   CN Tower, King and Spadina, Railway Lands, Har...              0.000   
10                                Caledonia-Fairbanks              0.000   
11              Canada Post Gateway Processing Centre              0.000   
12                                          Cedarbrae              0.000   
13                                 Central Bay Street              0.000   
14                                           Christie              0.000   
15                               Church and Wellesley              0.000   
16            Clarks Corners, Tam O'Shanter, Sullivan              0.000   
17    Cliffside, Cliffcrest, Scarborough Village West              0.000   
18                     Commerce Court, Victoria Hotel              0.000   
19                                         Davisville              0.000   
20                                   Davisville North              0.000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn              0.000   
22                                          Don Mills              0.000   
23  Dorset Park, Wexford Heights, Scarborough Town...              0.000   
24                                          Downsview              0.000   
25                       Dufferin, Dovercourt Village              0.000   
26      East Toronto, Broadview North (Old East York)              0.000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...              0.000   
28                       Fairview, Henry Farm, Oriole              0.000   
29             First Canadian Place, Underground city              0.000   
30    Forest Hill North & West, Forest Hill Road Park              0.000   
31                           Garden District, Ryerson              0.000   
32                                          Glencairn              0.000   
33                    Golden Mile, Clairlea, Oakridge              0.000   
34                  Guildwood, Morningside, West Hill              0.000   
35  Harbourfront East, Union Station, Toronto Islands              0.000   
36                      High Park, The Junction South              0.000   
37                                  Hillcrest Village              0.000   
38                                      Humber Summit              0.000   
39                                   Humberlea, Emery              0.000   
40                                 Humewood-Cedarvale              0.000   
41                     India Bazaar, The Beaches West              0.000   
42        Kennedy Park, Ionview, East Birchmount Park              0.000   
43          Kensington Market, Chinatown, Grange Park              0.000   
44  Kingsview Village, St. Phillips, Martin Grove ...              0.000   
45                   Lawrence Manor, Lawrence Heights              0.100   
46                                      Lawrence Park              0.000   
47                                            Leaside              0.000   
48                           Little Portugal, Trinity              0.000   
49                                     Malvern, Rouge              0.000   
50  Milliken, Agincourt North, Steeles East, L'Amo...              0.000   
51  Mimico NW, The Queensway West, South of Bloor,...              0.000   
52          

In [30]:
toronto_grouped.shape

(95, 273)

Here I show the top 5 venues per neighbourhood

In [31]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4          Accessories Store  0.00


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.25
1        Pharmacy  0.12
2             Pub  0.12
3  Sandwich Place  0.12
4             Gym  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
                       venue  freq
0                       Bank  0.09
1                Coffee Shop  0.09
2              Shopping Mall  0.04
3  Middle Eastern Restaurant  0.04
4          Mobile Phone Shop  0.04


----Bayview Village----
                 venue  freq
0                 Café  0.25
1                 Bank  0.25
2  Japanese Restaurant  0.25
3   Chinese Restaurant  0.25
4    Accessories Store  0.00


----Bedford Park, Lawrence Manor East----
                     venue  freq
0       Italian Restaurant  0.11
1              Co

Here I create a dataframe that shows the most common first 10 venue types per neighbourhood

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                Coffee Shop        Sandwich Place              Pharmacy   
2                       Bank         Shopping Mall        Sandwich Place   
3         Chinese Restaurant                  Bank   Japanese Restaurant   
4                Coffee Shop            Restaurant        Sandwich Place   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0   Dumpling Restaurant   Distribution Center                      Dog Run   
1                  Pool                   Pub                          Gym   
2            Restaurant           Supermarket               Ice Cream Shop   
3  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
4             Juice Bar            Hobby Shop                     Pharmacy   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                 Donut Shop              Drugstore  
1      Airport Terminal               Dessert Shop     Falafel Restaurant  
2      Sushi Restaurant  Middle Eastern Restaurant      Mobile Phone Shop  
3   Dumpling Restaurant                  Drugstore     Dim Sum Restaurant  
4           Pizza Place                        Pub                   Café

Here I apply the k-mean clustering technique to group the neighbourhoods depending on their venues 

In [34]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
print(kmeans.labels_)
len(kmeans.labels_)

[1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 4 1 1 1 1 1 1
 1 1 2 1 1 1 1 1 1 1 1 1 1 0 1 4 1 4 1 1 1 2 1 1 1 1 1 1 4 1 3 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0]


95

In [35]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = wiki_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.dropna(inplace = True)
toronto_merged.reset_index(drop = True, inplace = True)

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels   1st Most Common Venue  \
0  43.753259 -79.329656             1.0       Food & Drink Shop   
1  43.725882 -79.315572             1.0            Hockey Arena   
2  43.654260 -79.360636             1.0             Coffee Shop   
3  43.718518 -79.464763             1.0  Furniture / Home Store   
4  43.662301 -79.389494             1.0             Coffee Shop   

  2nd Most Common Venue  3rd Most Common Venue       4th Most Common Venue  \
0                  Park      Convenience Store                 Yoga Studio   
1           Coffee Shop  Portuguese Restaurant  Financial or Legal Service   
2                  Park                 Bakery                         Pub   
3        Clothing Store  Vietnamese Restaurant                    Boutique   
4                 Diner     College Auditorium                         Bar   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0                 Diner        Discount Store          Distribution Center   
1     French Restaurant           Pizza Place  Eastern European Restaurant   
2        Breakfast Spot                  Café                      Theater   
3             Gift Shop     Accessories Store                  Coffee Shop   
4              Beer Bar         Smoothie Shop               Sandwich Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run      Doner Restaurant             Donut Shop  
1   Dumpling Restaurant             Drugstore      Electronics Store  
2            Beer Store            Shoe Store             Restaurant  
3           Event Space          Concert Hall             Comic Shop  
4         Burrito Place                  Café  Portuguese Restaurant

Here I create a map that shows the result of the grouping (5 clusters)

In [36]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    #print(cluster)
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Here I analyse the clusters

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
19                                Caledonia-Fairbanks             0.0   
49                            Willowdale, Newtonbrook             0.0   
61                                             Weston             0.0   
63                                    York Mills West             0.0   
82  Milliken, Agincourt North, Steeles East, L'Amo...             0.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
19                  Park         Women's Store                  Pool   
49                  Park           Yoga Studio             Drugstore   
61                  Park           Yoga Studio             Drugstore   
63                  Park     Convenience Store           Yoga Studio   
82                  Park            Playground           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
19            Donut Shop    Dim Sum Restaurant                 Diner   
49                 Diner        Discount Store   Distribution Center   
61                 Diner        Discount Store   Distribution Center   
63             Drugstore                 Diner        Discount Store   
82            Donut Shop    Dim Sum Restaurant                 Diner   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
19        Discount Store   Distribution Center               Dog Run   
49               Dog Run      Doner Restaurant            Donut Shop   
61               Dog Run      Doner Restaurant            Donut Shop   
63   Distribution Center               Dog Run      Doner Restaurant   
82        Discount Store   Distribution Center               Dog Run   

   10th Most Common Venue  
19       Doner Restaurant  
49    Dumpling Restaurant  
61    Dumpling Restaurant  
63             Donut Shop  
82       Doner Restaurant

First cluster: it seems to group neighbourhoods with access to green area and relaxing activities (park, yoga studio, playground), good for a nature lover

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
0                                           Parkwoods             1.0   
1                                    Victoria Village             1.0   
2                           Regent Park, Harbourfront             1.0   
3                    Lawrence Manor, Lawrence Heights             1.0   
4         Queen's Park, Ontario Provincial Government             1.0   
5                                      Malvern, Rouge             1.0   
6                                           Don Mills             1.0   
7                     Parkview Hill, Woodbine Gardens             1.0   
8                            Garden District, Ryerson             1.0   
9                                           Glencairn             1.0   
11                                          Don Mills             1.0   
12                                   Woodbine Heights             1.0   
13                                     St. James Town             1.0   
14                                 Humewood-Cedarvale             1.0   
15  Eringate, Bloordale Gardens, Old Burnhamthorpe...             1.0   
16                  Guildwood, Morningside, West Hill             1.0   
17                                        The Beaches             1.0   
18                                        Berczy Park             1.0   
20                                             Woburn             1.0   
21                                            Leaside             1.0   
22                                 Central Bay Street             1.0   
23                                           Christie             1.0   
24                                          Cedarbrae             1.0   
25                                  Hillcrest Village             1.0   
26    Bathurst Manor, Wilson Heights, Downsview North             1.0   
27                                   Thorncliffe Park             1.0   
28                           Richmond, Adelaide, King             1.0   
29                       Dufferin, Dovercourt Village             1.0   
30                                Scarborough Village             1.0   
31                       Fairview, Henry Farm, Oriole             1.0   
32                    Northwood Park, York University             1.0   
33      East Toronto, Broadview North (Old East York)             1.0   
34  Harbourfront East, Union Station, Toronto Islands             1.0   
35                           Little Portugal, Trinity             1.0   
36        Kennedy Park, Ionview, East Birchmount Park             1.0   
37                                    Bayview Village             1.0   
38                                          Downsview             1.0   
39                       The Danforth West, Riverdale             1.0   
40           Toronto Dominion Centre, Design Exchange             1.0   
41       Brockton, Parkdale Village, Exhibition Place             1.0   
42                    Golden Mile, Clairlea, Oakridge             1.0   
43                                          Downsview             1.0   
44                     India Bazaar, The Beaches West             1.0   
45                     Commerce Court, Victoria Hotel             1.0   
47                                      Humber Summit             1.0   
48    Cliffside, Cliffcrest, Scarborough Village West             1.0   
50                                          Downsview             1.0   
51                                    Studio District             1.0   
52                  Bedford Park, Lawrence Manor East             1.0   
53   Del Ray, Mount Dennis, Keelsdale and Silverthorn             1.0   
55                        Birch Cliff, Cliffside West             1.0   
56                        Willowdale, Willowdale East             1.0   
57                                          Downsview             1.0   
58                                      Lawrence Park             1.0   
59                       

Second cluster: the most numerous, it seems to group neighbourhoods with very different services

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
54  North York             2.0          Food Service        Baseball Field   
97   Etobicoke             2.0        Baseball Field           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
54           Yoga Studio   Distribution Center               Dog Run   
97             Drugstore        Discount Store   Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
54      Doner Restaurant            Donut Shop             Drugstore   
97               Dog Run      Doner Restaurant            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
54   Dumpling Restaurant                  Diner  
97   Dumpling Restaurant                  Field

Third cluster: just two elements, they seem to be quite out of the city

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighbourhood  Cluster Labels  \
10  Rouge Hill, Port Union, Highland Creek             3.0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                   Bar           Yoga Studio             Drugstore   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
10        Discount Store   Distribution Center               Dog Run   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10      Doner Restaurant            Donut Shop   Dumpling Restaurant   

   10th Most Common Venue  
10                  Field

Fourth cluster: just 1 element, it seems to offer just perypherical activities

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Fifth cluster: mostly neighbourhoods next to a park